### PR TITLE
graph: always use the on-disk mtime of generator rules

### PR DESCRIPTION
**Pull request by mathstuf**
Generator rules are likely to have been run outside of ninja, so the
build log should not be trusted for detecting whether or not to rerun
the rule.

---
Observed with CMake projects which exhibit behavior that `ninja` reruns `cmake` right after running `cmake` manually. Maybe it is better to just not add generator rules to the build log at all? I don't know of the implications of that.

See https://gitlab.kitware.com/cmake/cmake/issues/15830 for some context.

### CODE REVIEW

**Comment by fkaelberer** [comment]
So this should Fix #1376, I guess.

**Comment by mathstuf** [comment]
Yes, it should. I suspect there are likely other duplicates of the issue around.

**Comment by Siron777** [comment]
I have tested the patch and I do not observe anymore the issue #1376.

File: src/graph.cc [hunk 1]
```diff
@@ -252,14 +252,21 @@ bool DependencyScan::RecomputeOutputDirty(Edge* edge,
         EXPLAIN("command line changed for %s", output->path().c_str());
         return true;
       }
-      if (most_recent_input && entry->mtime < most_recent_input->mtime()) {
+      // Compute the effective mtime for the rule.
+      TimeStamp effective_mtime = entry->mtime;
+      if (generator) {
+        // For generator rules, always use the on-disk mtime since it is likely
+        // to have been regenerated outside of ninja.
+        effective_mtime = output->mtime();
```
Comment (jhasse) [inline src/graph.cc]: How can you be sure that `mtime()` doesn't return `-1`?

Comment (mathstuf) [reply src/graph.cc]: Hmm. It is used above (line 220). I'll test out adding `&& output->status_known()` to the conditional guarding this.

Comment (mathstuf) [reply src/graph.cc]: Still works with that conditional. I don't know exactly how the status wouldn't be known at this point, but it doesn't happen in typical CMake usage patterns at least.

**Comment by rubenvb** [comment]
This seems to fix the issue for me.

**Comment by jhuels** [comment]
This fixes the issue for `generator` rule for me, but I don't think CMake provides a way to mark custom commands as `generator`.  I'm having an identical issue with a code generator where the disc time of the output is newer because it was run outside of ninja, but ninja_log is stale.

Do you know of a way around this, or should I request a feature for CMake to set custom commands as `generator`?

**Comment by mathstuf** [comment]
Please open an issue over there (rather than muddying the discussion here), but include your example custom command. Ninja's rules on the `generator` rule variable doesn't really make sense with custom commands.

**Comment by jhasse** [comment]
> Do you know of a way around this

Update the ninja_log in your custom command.

**Comment by mathstuf** [comment]
> Update the ninja_log in your custom command.

I don't know that this is going to be easy. Ninja itself uses the tmpfile-and-rename style, so any command doing this would need to coordinate with ninja on what's going on. This just sounds 100% flaky to me.

**Comment by jhasse** [comment]
> Ninja itself uses the tmpfile-and-rename style, so any command doing this would need to coordinate with ninja on what's going on.

But this shouldn't matter because we're talking about the case where the generator is run outside of ninja, right?

**Comment by mathstuf** [comment]
How is the tool supposed to know if it is inside or outside ninja? Plus, the question asked was for a way to claim `generator = 1` in order to access the behavior introduced here. Maybe some other flag should be added that `generator = 1` should imply?

**Comment by jhasse** [comment]
> How is the tool supposed to know if it is inside or outside ninja?

Pass itself a flag in its build rule for example.

**Comment by mathstuf** [comment]
Maybe if the deplog format were documented and guaranteed to be stable, that route could be taken. Until then, is this a sufficient solution to the problem being seen?

**Comment by jhasse** [comment]
I would rather add a tool to Ninja (e.g. `ninja -t updatecache`?) or add the functionality to an existing one. What about `ninja -t recompact`? Or the [new `cleandead` tool](https://github.com/ninja-build/ninja/pull/1432)? Running the latter on reconfigures is a good idea anyway, as it might catch bugs by deleting files hanging around from previous builds.

**Comment by mathstuf** [comment]
Hmm. That could work. A pipeline like `ninja -t cleandead && ninja -t recompact && ninja -t restat` maybe?

**Comment by jhasse** [comment]
Exactly! `restat` also sounds like a good name.

**Comment by mathstuf** [comment]
Strawman MR filed for CMake here: https://gitlab.kitware.com/cmake/cmake/merge_requests/3316

**Comment by jhasse** [comment]
Closing in favor of #1685.

**Pull request closed without merging**